extensions/amp-list/0.1/test/test-amp-list-load-more.js

```diff
@@ -0,0 +1,125 @@
+
+import {AmpDocService} from '../../../../src/service/ampdoc-impl';
+import {AmpList} from '../amp-list';
+import {Services} from '../../../../src/services';
+import {isExperimentOn, toggleExperiment} from '../../../../src/experiments';
+
+
+describes.realWin('amp-list component', {
+  amp: {
+    ampdoc: 'single',
+    extensions: ['amp-list'],
+  },
+}, env => {
+  let win, doc, ampdoc, sandbox;
+  let element, list;
+  let templates;
+
+  beforeEach(() => {
+    win = env.win;
+    doc = win.document;
+    ampdoc = env.ampdoc;
+    sandbox = env.sandbox;
+
+    templates = {
+      findAndSetHtmlForTemplate: sandbox.stub(),
+      findAndRenderTemplate: sandbox.stub(),
+      findAndRenderTemplateArray: sandbox.stub(),
+    };
+    sandbox.stub(Services, 'templatesFor').returns(templates);
+    sandbox.stub(AmpDocService.prototype, 'getAmpDoc').returns(ampdoc);
+
+    toggleExperiment(win, 'amp-list-load-more', true);
+  });
+
+  describe('manual', () => {
+
+    beforeEach(() => {
+      expect(isExperimentOn(win, 'amp-list-load-more')).to.be.true;
+
+      element = doc.createElement('amp-list');
+      list = new AmpList(element);
+
+      sandbox.stub(list, 'getAmpDoc').returns(ampdoc);
+      sandbox.stub(list, 'getFallback').returns(null);
+
+      sandbox.stub(list, 'mutateElement').callsFake(mutator => {
+        mutator();
+        return Promise.resolve();
+      });
+
+      sandbox.stub(list, 'measureElement').callsFake(measurer => {
+        measurer();
+        return Promise.resolve();
+      });
+
+      sandbox.stub(list, 'measureMutateElement').callsFake(
+          (measurer, mutator) => {
+            measurer();
+            mutator();
+            return Promise.resolve();
+          });
+
+      element.setAttribute('src', '/list/infinite-scroll?items=2&left=1');
+      element.setAttribute('load-more', 'manual');
+      element.setAttribute('layout', 'fixed');
+      element.setAttribute('width', '50');
+      element.setAttribute('height', '10');
+
+      element.style.height = '10px';
+      doc.body.appendChild(element);
+
+      sandbox.stub(list, 'getOverflowElement').returns(null);
+      sandbox.stub(list, 'fetchList_').returns(Promise.resolve());
+      list.element.changeSize = () => {};
+      list.buildCallback();
+    });
+
+    it('should create load-more elements after layout callback', async() => {
+      sandbox.stub(list, 'getPlaceholder').returns(null);
+      await list.layoutCallback();
+      expect(list.element.querySelector('[load-more-button]')).to.be.ok;
+      expect(list.element.querySelector('[load-more-failed]')).to.be.ok;
+      expect(list.element.querySelector('[load-more-end]')).to.be.null;
+    });
+
+    it('should hide load-more elements at layout', async() => {
+      sandbox.stub(list, 'getPlaceholder').returns(null);
+      await list.layoutCallback();
+      const button = list.element.querySelector('[load-more-button]');
+      const buttonStyles = win.getComputedStyle(button);
```

Comment:
I think you can also do:

```js
expect(buttonStyles).to.include({
  'display': 'block',
  'visibility': 'hidden',
});
```

This is a bit nicer in that it will tell you if one or both are wrong. With the existing `expect`s if the button `display` is wrong, the test will just fail there, without you knowing the `visibility` is wrong as well. So you may make some changes to fix `display`, only to find out you need to fix `visibility` as well.

To extend this train of thought, ideally you would have two separate test cases, one for the load-more-button and one for the load-more-failed, so you can tell if they are wrong independently (as some people say, one test should test a single thing). No need to do that here, since there is some cost to a separate test case here.

Reply:
This isn't that bad at all and quite worth doing.

Reply:
Done.